clang/lib/Sema/Metafunctions.cpp

```diff
@@ -4595,20 +4622,83 @@ bool reflect_invoke(APValue &Result, Sema &S, EvalFn Evaluator,
   ExprResult ER;
   {
     EnterExpressionEvaluationContext Context(
-            S, Sema::ExpressionEvaluationContext::ConstantEvaluated);
-    if (auto *DRE = dyn_cast<DeclRefExpr>(FnRefExpr);
-        DRE && dyn_cast<CXXConstructorDecl>(DRE->getDecl())) {
-      auto *CtorD = cast<CXXConstructorDecl>(DRE->getDecl());
+        S, Sema::ExpressionEvaluationContext::ConstantEvaluated);
 
+    auto *DRE = dyn_cast<DeclRefExpr>(FnRefExpr);
+    if (DRE && dyn_cast<CXXConstructorDecl>(DRE->getDecl())) {
+      auto *CtorD = cast<CXXConstructorDecl>(DRE->getDecl());
       ER = S.BuildCXXConstructExpr(
-            Range.getBegin(), QualType(CtorD->getParent()->getTypeForDecl(), 0),
-            CtorD, false, ArgExprs, false, false, false, false,
-            CXXConstructionKind::Complete, Range);
+          Range.getBegin(), QualType(CtorD->getParent()->getTypeForDecl(), 0),
+          CtorD, false, ArgExprs, false, false, false, false,
+          CXXConstructionKind::Complete, Range);
     } else {
-      ER = S.ActOnCallExpr(S.getCurScope(), FnRefExpr, Range.getBegin(),
-                           ArgExprs, Range.getEnd(), /*ExecConfig=*/nullptr);
+      auto *FnExpr = FnRefExpr;
+      auto FnArgExprs = ArgExprs;
+
+      if (DRE && is_nonstatic_member_function(DRE->getDecl())) {
+        auto *MD = cast<CXXMethodDecl>(DRE->getDecl());
+
+        if (ArgExprs.size() < 1) {
+          // need to have object as a first argument
+          return Diagnoser(Range.getBegin(),
+                           diag::metafn_first_argument_is_not_object)
+                 << Range;
+        }
+
+        auto ObjExpr = ArgExprs[0];
+        auto ObjType = ObjExpr->getType();
+
+        if (ObjType->isPointerType()) {
+          ObjType = ObjType->getPointeeType();
+        }
+
+        if (!ObjType->getAsCXXRecordDecl()) {
+          // first argument is not an object
+          return Diagnoser(Range.getBegin(),
+                           diag::metafn_first_argument_is_not_object)
+                 << Range;
+        }
+
+        // check that method belongs to class
+        if (ObjType->getAsCXXRecordDecl()->getDeclName() !=
+            MD->getParent()->getDeclName()) {
+          return Diagnoser(Range.getBegin(),
+                           diag::metafn_function_is_not_member_of_object)
+                 << Range;
+        }
+
+        if (MD->getReturnType()->isVoidType()) {
+          // void return type is not supported
+          return Diagnoser(Range.getBegin(), diag::metafn_function_returns_void)
+                 << Range;
+        }
+
+        SourceLocation ObjLoc = ObjExpr->getExprLoc();
+        UnqualifiedId Name;
+        Name.setIdentifier(MD->getIdentifier(), ObjLoc);
```

Comment:
Hmmm this might be a problem too. If we do lookup, then we might get things other than `MD` - for instance, perhaps the name will lookup an overload set, and then chooses a different overload (with the same name as `MD`) to execute. `reflect_invoke` should not do lookup or overload resolution.

This is a problem that we run into with splicing (e.g., `obj.[:fn():]`) as well, and I had to hack a different interface on the member-access-expr machinery to make it work.
https://github.com/bloomberg/clang-p2996/blob/p2996/clang/lib/Sema/SemaExprMember.cpp#L1900-L1906

You could try synthesizing a `DeclRefExpr`, wrapping it in a `CXXSpliceExpr`, and using that to call to the variant linked above. I'll have to think of whether there's a more efficient way, but I think that should work for now.

Reply:
I tried to implement this approach and it seems to work but I am not sure on it, might have missed something

Please look [8d32747](https://github.com/bloomberg/clang-p2996/pull/79/commits/8d327476102a1c488297de2e9879db845caa279e)